### PR TITLE
Implement topic distribution assessment

### LINKED ISSUE #1546: Make a topicDistribution research
We need to implement a new topicDistribution research, which takes into consideration not only focus keyword but also synonyms.

### PR DESCRIPTION
**Pull request by nataliashitova**
## Summary

This PR can be summarized in the following changelog entry:

* Implements topic distribution assessment, i.e., distribution of the focus keyword and its synonyms in the text.
* Adds markings to the density assessment, independently on whether synonyms are supplied or not.

## Relevant technical choices:

* Until the synonym interface is implemented, synonyms are parsed from the focus keyword field, comma-separated. Right now there is no difference between Free and Premium. The synonym interface should take care of making the synonym field available for Premium only.

## Test instructions

This PR can be tested by following these steps:

### Browserified or beta
* Enter a text of more than 200 words, enter a keyword that occurs in the text at least twice. The keyword density and keyword distribution assessments should return the result in terms of keyword matches (i.e., there should be no mention of synonyms). Check if markings are ok.
* Enter a synonym, separated from the keyword by a comma. The assessments should return the result in terms of topic (i.e., there should be feedback that mentions both keyword and its synonyms). Check markings.
* Try to see if the boundaries for keyword alone or keyword+synonyms are different.

Fixes #1546

### CODE REVIEW

File: spec/assessments/largestKeywordDistanceAssessmentSpec.js [hunk 1]
```diff
@@ -7,23 +7,23 @@ const Mark = require( "../../js/values/Mark.js" );
...
+	it( "returns an okay score when the largest keyword distance is between 40 and 50%", function() {
 		let mockPaper = new Paper( "string with the keyword", { keyword: "keyword" } );
 		let assessment = keywordDistanceAssessment.getResult( mockPaper, Factory.buildMockResearcher( 45 ), i18n );
 
 		expect( assessment.getScore() ).toEqual( 6 );
 		expect( assessment.getText() ).toEqual( "Some parts of your text do not contain the keyword. Try to distribute the keyword more evenly." );
 	} );
 
-	it( "returns an okay score when the largest keyword distance is less than 30%", function() {
+	it( "returns an okay score when the largest keyword distance is less than 40%", function() {
```
Comment (manuelaugustin) [inline spec/assessments/largestKeywordDistanceAssessmentSpec.js]: Apparently a pre-existing mistake: "returns an okay score" should be "returns a good score".

File: spec/assessments/largestKeywordDistanceAssessmentSpec.js [hunk 2]
```diff
@@ -32,6 +32,35 @@ describe( "An assessment to check the largest percentage of text in which no key
 	} );
 } );
 
+describe( "An assessment to check the largest percentage of text in which no keyword or synonyms occured", function() {
+	it( "returns a bad score when the largest keyword distance is between more than 40%", function() {
```
Comment (manuelaugustin) [inline spec/assessments/largestKeywordDistanceAssessmentSpec.js]: `is more than 40%` (remove `between`)

**Comment by manuelaugustin** [comment]
CR 🚧

* See comments for feedback.

* In `indicesSpecs.js`, could you add specs for `getIndicesByWordListSorted`?

* This is a mistake I had introduced previously, but could change the incorrect description `additionally runs assessments that require an url` in SEOAssessorSpec to `additionally runs assessments that require a long enough text and two keyword occurrences` (same as in cornerstone/SEOAssessorSpec.js)?

File: spec/assessments/largestKeywordDistanceAssessmentSpec.js (hunk 1 above)
Comment (nataliashitova) [reply spec/assessments/largestKeywordDistanceAssessmentSpec.js]: Fixed.

File: spec/assessments/largestKeywordDistanceAssessmentSpec.js (hunk 2 above)
Comment (nataliashitova) [reply spec/assessments/largestKeywordDistanceAssessmentSpec.js]: Fixed.

File: spec/assessments/largestKeywordDistanceAssessmentSpec.js [hunk 3]
```diff
@@ -32,6 +32,35 @@ describe( "An assessment to check the largest percentage of text in which no key
...
+	it( "returns an okay score when the largest keyword distance is between 30 and 40%", function() {
+		// todo: change keyword definition
+		let mockPaper = new Paper( "string with the keyword", { keyword: "keyword, synonym, synonyms" } );
+		let assessment = keywordDistanceAssessment.getResult( mockPaper, Factory.buildMockResearcher( 35 ), i18n );
+
+		expect( assessment.getScore() ).toEqual( 6 );
+		expect( assessment.getText() ).toEqual( "Some parts of your text do not contain the keyword or its synonyms. Try to distribute them more evenly." );
+	} );
+
+	it( "returns an okay score when the largest keyword distance is less than 30%", function() {
```
[the remark this answers is not in the document]
Comment (nataliashitova) [reply spec/assessments/largestKeywordDistanceAssessmentSpec.js]: Fixed.

File: src/stringProcessing/indices.js [hunk 4]
```diff
@@ -91,9 +91,44 @@ var filterIndices = function( indices ) {
...
+
+	forEach( words, function( word ) {
+		word = stripSpaces( word );
+		if ( ! matchWordInSentence( word, text ) ) {
+			return matchedWords;
+		}
+		matchedWords = matchedWords.concat( getIndicesByWord( word, text ) );
+	} );
+
+	let matchedWordsArrayed = Array.from( matchedWords );
```
[the remark this answers is not in the document]
Comment (nataliashitova) [reply src/stringProcessing/indices.js]: Fixed.

**Comment by manuelaugustin** [comment]
CR & acceptance done 👍

**Pull request merged**